Improved GEV documentation

adutfoy opened the pull request:
I improved the documentation of the GEV developments: 

- ProfileLikelihoodResult
- GeneralizedExtremeValueFactory
- HypothesisTest::LikelihoodRatioTest
- TimeVaryingResult
- MaximumLikelihoodFactory

jschueller commented on lib/src/Uncertainty/Distribution/GeneralizedExtremeValueFactory.cxx:
```diff
@@ -973,16 +973,15 @@ ProfileLikelihoodResult GeneralizedExtremeValueFactory::buildReturnLevelProfileL
   const Matrix covZm = (dzm.transpose() * Vn * dzm);
   Normal parameterDistribution(optimalParameter, CovarianceMatrix(covZm.getImplementation()));
-  parameterDistribution.setDescription({"zm", "sigma", "xi"});
+  parameterDistribution.setDescription({buildReturnLevelProfileLikelihoodEstimator});
```
probably a mistake here

adutfoy replied:
ok You are right..... Thanks!

jschueller commented:
did you have a look at python/doc/theory/data_analysis/tail_dependence.rst ?

adutfoy commented:
Coming tasks: review of the files about the tail dependence coefficients and all the examples.

jschueller commented on python/doc/examples/data_analysis/distribution_fitting/plot_estimate_gev_fremantle.py:
```diff
@@ -3,164 +3,218 @@
 print(data[:5])
+cloud = ot.Cloud(data[:, :2])
+cloud.setColor("red")
+graph.add(cloud)
```
graph is not set

adutfoy replied:
yes.... I modifiy it.

jschueller commented on python/doc/examples/data_analysis/distribution_fitting/plot_estimate_gev_fremantle.py:
```diff
@@ -82,7 +82,7 @@
 
 # %%
-# **Stationary GEV modeling via the profile log-likelihood function**
+# **Stationary GEV mocoefficientsdeling via the profile log-likelihood function**
```
oups

adutfoy replied:
Ok! Do you understand the error message : 
Warning, treated as error:
/home/maison/openturns-anne/build_doc_gev/python/src/site-packages/openturns/dist_bundle3.py:docstring of openturns.model_copula.DistributionImplementation.drawLowerExtremalDependenceFunction:7:Libellé dupliqué pour l'équation lowerExtremDpceFunc, autre instance dans user_manual/_generated/openturns.SquaredNormal

adutfoy commented:
@jschueller 
Help! I really do not understand what happens here... Can you help me?
Thanks!
Anne

adutfoy commented:
@jschueller 
We got it!!!! a s after `openturns.BasisCollection`....

jschueller commented:
good, I was only narrowing it down to buildTimeVarying

jschueller commented on lib/src/Uncertainty/Distribution/ProfileLikelihoodResult.cxx:
```diff
@@ -35,17 +35,26 @@ static const Factory<ProfileLikelihoodResult> Factory_ProfileLikelihoodResult;
+  , xMax_(xMax)
+{
+  // Nothing to do
+  std::cout << "parameter=" << parameter << ", xMin=" << xMin << ", xMax=" << xMax << std::endl;
```
you have some debug cout left

jschueller commented on lib/src/Uncertainty/Distribution/GeneralizedExtremeValueFactory.cxx:
```diff
@@ -975,7 +1320,23 @@ ProfileLikelihoodResult GeneralizedExtremeValueFactory::buildReturnLevelProfileL
+  const Scalar xiMax =  5 * sigma / (mu - zMin);
+  const Scalar zmMin = mu + sigma * std::expm1(xiMin * logLog1pM) / xiMin;
+  const Scalar zmMax = mu + sigma * std::expm1(xiMax * logLog1pM) / xiMax;
+  std::cout << "mu=" << mu << ", sigma=" << sigma << ", xi=" << xi << ", zm=" << zm << ", zMin=" << zMin << ", zMax=" << zMax << ", xiMin=" << xiMin << ", xiMax=" << xiMax << ", zmMin=" << zmMin << ", zmMax=" << zmMax << std::endl;
```
here too

adutfoy commented:
Hi
My PR is ready to be reviewed!
I modified the examples and added the comments of the results:

- /python/doc/examples/data_analysis/distribution_fitting/plot_estimate_gev_fremantle.py
-/python/doc/examples/data_analysis/distribution_fitting/ plot_estimate_gev_pirie.py
-/python/doc/examples/data_analysis/distribution_fitting/ plot_estimate_gev_racetime.py
-/python/doc/examples/data_analysis/distribution_fitting/ plot_estimate_gev_venice.py
- /python/doc/examples/data_analysis/estimate_dependency_and_copulas/plot_estimate_dependence_wavesurge.py
- /python/doc/examples/data_analysis/estimate_dependency_and_copulas/plot_estimate_dependence_wind.py

The examples provide the same results as Coles.

I modified the documentation of:

- GeneralizedExtremeValueFactory
- GeneralizedExtremeValueValidation
- ProfileLikelihoodResult
- HypothesisTest::LikelihoodRatioTest
- TimeVaryingResult
- MaximumLikelihoodFactory

@jschueller @josephmure @mbaudin47 : could you review this PR???

Thanks!
Anne

jschueller commented on lib/src/Uncertainty/Distribution/openturns/GeneralizedExtremeValueFactory.hxx:
```diff
@@ -60,25 +61,32 @@ public:
+                                    const String & initializationMethod = ResourceMap::GetAsString("GeneralizedExtremeValueFactory-InitializationMethod"),
+                                    const String & normalizationMethod = ResourceMap::GetAsString("GeneralizedExtremeValueFactory-NormalizationMethod")) const;
+#endif
   /** Return level */
```
we should drop that part if its not enabled, and also CovariatesResult which seems unused

adutfoy replied:
yes! we are going to remove the code related to covariates.

adutfoy replied:
The lines you mention are used! There are strings that enable to choose the initialization method and the initial point!

jschueller replied:
I was referring to the buildCovariates method

jschueller commented on lib/src/Uncertainty/Distribution/openturns/GeneralizedExtremeValueFactory.hxx:
```diff
@@ -60,25 +61,32 @@ public:
-
+                                     const Function & inverseLinkFunction = Function(),
+                                     const String & initializationMethod = ResourceMap::GetAsString("GeneralizedExtremeValueFactory-InitializationMethod"),
+                                     const String & normalizationMethod = ResourceMap::GetAsString("GeneralizedExtremeValueFactory-NormalizationMethod")) const;
```
What about removing initializationMethod & normalizationMethod arguments ?
this function already has enough and these can be set in the ResourceMap anyways

adutfoy replied:
It was asked by our internal users, in the spirit of the R package ismev. The use of ResourceMap is not straightforward for most users.

mbaudin47 commented on lib/src/Uncertainty/Distribution/GeneralizedExtremeValueFactory.cxx:
```diff
@@ -120,124 +138,91 @@ class GeneralizedExtremeValueLikelihoodEvaluation : public EvaluationImplementat
+
+        for (UnsignedInteger k = 0; k < r_ - 1; ++ k)
+        {
+          const Scalar yik = (sample_(i, k) - mu) / sigma;
+          ll += -yik;
```
```suggestion
          ll += -(sample_(i, k) - mu) / sigma;
```

mbaudin47 commented on lib/src/Uncertainty/Distribution/GeneralizedExtremeValueFactory.cxx:
```diff
@@ -120,124 +138,91 @@ class GeneralizedExtremeValueLikelihoodEvaluation : public EvaluationImplementat
+        if (c1 <= SpecFunc::Precision - 1.0) // can be slightly off
+        {
+          ll += -std::log(SpecFunc::MaxScalar);
+          // std::cerr << "i=" << i << ", c1<-1:" << c1 << ", xi=" << xi << ", yir=" << yir << ", ll=" << ll << std::endl;
```
This line is an unnecessary comment:
```suggestion
```

mbaudin47 commented on lib/src/Uncertainty/Distribution/GeneralizedExtremeValueFactory.cxx:
```diff
@@ -120,124 +138,91 @@ class GeneralizedExtremeValueLikelihoodEvaluation : public EvaluationImplementat
+          if (c2 <= SpecFunc::Precision - 1.0) // can be slightly off
+          {
+            ll += -std::log(SpecFunc::MaxScalar);
+            // std::cerr << "i=" << i << ", c2<-1:" << c1 << ", xi=" << xi << ", yik=" << yik << ", ll=" << ll << std::endl;
```
```suggestion
```

mbaudin47 commented on lib/src/Uncertainty/Distribution/GeneralizedExtremeValueFactory.cxx:
```diff
@@ -120,124 +138,91 @@ class GeneralizedExtremeValueLikelihoodEvaluation : public EvaluationImplementat
+
+        for (UnsignedInteger k = 0; k < r_ - 1; ++ k)
+        {
+          const Scalar yik = (sample_(i, k) - mu) / sigma;
+          const Scalar c2 = xi * yik;
```
```suggestion
          const Scalar c2 = xi * (sample_(i, k) - mu) / sigma;
```

mbaudin47 commented on lib/src/Uncertainty/Distribution/GeneralizedExtremeValueFactory.cxx:
```diff
@@ -289,15 +274,24 @@ class GeneralizedExtremeValueProfileLikelihoodEvaluation : public EvaluationImpl
+        if (cv[1] <= 0.0)
+          x0[0] = zMin_;
+      }
+    // std::cerr << "Final x0=" << x0 << ", cv=" << constraint(x0) << std::endl;
```
```suggestion
```

mbaudin47 commented on lib/src/Uncertainty/Distribution/GeneralizedExtremeValueFactory.cxx:
```diff
@@ -585,80 +509,135 @@ class GeneralizedExtremeValueTimeVaryingLikelihoodEvaluation : public Evaluation
+      if (c1 <= SpecFunc::Precision - 1.0) // can be slightly off
+      {
+        ll += -std::log(SpecFunc::MaxScalar);
+        // std::cerr << "c1<-1:" << c1 << ", xi=" << xi << ", yi=" << yi << ", ll=" << ll << std::endl;
```
```suggestion
```

mbaudin47 commented on lib/src/Uncertainty/Distribution/GeneralizedExtremeValueFactory.cxx:
```diff
@@ -975,7 +976,37 @@ ProfileLikelihoodResult GeneralizedExtremeValueFactory::buildReturnLevelProfileL
+  // zmMin = mu + sigma * (exp(-xiMin * log(-log(1-1/m))) - 1) / xiMin
+  // zmMin = mu + sigma * (exp(-xiMax * log(-log(1-1/m))) - 1) / xiMax
+  // logLog1pM = log(-log(1-p)) p<=1/2, 1-p >= 1/2, log(1-p) >= -log(2), -log(1-p) <= log(2), log(-log(1-p)) <= log(log(2)), log(-log(1-p)) <= log(log(2)) < 0
+  /* Unfortunately this nice idea fails...
+    const Scalar zMin = sample.getMin()[0];
+    const Scalar zMax = sample.getMax()[0];
+    Scalar zmMin;
+    if (zMax > mu)
+    {
+    const Scalar xiMin = sigma / (mu - zMax);
+    zmMin = mu + sigma * std::expm1(-xiMin * logLog1pM) / xiMin;
+    }
+    else
+    // mu + sigma * (0 - 1) / (-inf) = mu
+    zmMin = mu;
+    Scalar zmMax;
+    if (zMin < mu)
+    {
+    const Scalar xiMax = sigma / (mu - zMin);
+    zmMax = mu + sigma * std::expm1(-xiMax * logLog1pM) / xiMax;
+    }
+    else
+    zmMax = SpecFunc::MaxScalar;
+    std::cerr << "mu=" << mu << ", sigma=" << sigma << ", zMin=" << zMin << ", zMax=" << zMax << ", zmMin=" << zmMin << ", zmMax=" << zmMax << std::endl;
+  */
```
```suggestion
```

mbaudin47 commented on lib/src/Uncertainty/Distribution/ProfileLikelihoodResult.cxx:
```diff
@@ -111,34 +122,68 @@ Interval ProfileLikelihoodResult::getParameterConfidenceInterval() const
+    ci = getParameterConfidenceInterval();
+  }
+  catch (const Exception &)
+  {
+    drawCI = false;
+  }
```
```suggestion
    drawCI = false;
```

mbaudin47 commented on lib/src/Uncertainty/Distribution/ProfileLikelihoodResult.cxx:
```diff
@@ -111,34 +122,68 @@ Interval ProfileLikelihoodResult::getParameterConfidenceInterval() const
+  Interval ci(xMin_, xMax_);
+  Bool drawCI = true;
+  try
+  {
+    ci = getParameterConfidenceInterval();
+  }
```
```suggestion
   ci = getParameterConfidenceInterval();
```

mbaudin47 commented on lib/src/Uncertainty/Distribution/ProfileLikelihoodResult.cxx:
```diff
@@ -111,34 +122,68 @@ Interval ProfileLikelihoodResult::getParameterConfidenceInterval() const
+      result = profileLikelihoodFunction_.draw(xMin, xMax);
+    }
   result.setLegends({"likelihood"});
+  // std::cerr << "xMin=" << xMin << ", x=" << x << ", xMax=" << xMax << std::endl;
```
```suggestion
```

mbaudin47 reviewed (commented):
I see some intermediate debug code still in the source: I suggest to removed this.

mbaudin47 commented on python/doc/examples/data_analysis/distribution_fitting/plot_estimate_gev_fremantle.py:
```diff
@@ -1,167 +1,470 @@
-# Now formulate a non-stationary hypothesis for our parameters
-# First we need the grid of time values (in years)
-mesh = ot.Mesh(data[:, 0])
+# **Non stationary GEV modeling via the log-likelihood function**
```
```suggestion
# Non stationary GEV modeling via the log-likelihood function
# -----------------------------------------------------------
```

adutfoy replied:
idem

mbaudin47 commented on python/doc/examples/data_analysis/distribution_fitting/plot_estimate_gev_fremantle.py:
```diff
@@ -1,167 +1,470 @@
 # %%
-# Now estimate the parameters with the profile likelihood
-result2 = factory.buildMethodOfProfileLikelihoodMaximizationEstimator(sample)
+# **Stationary GEV modeling via the profile log-likelihood function**
```
```suggestion
# Stationary GEV modeling via the profile log-likelihood function
# ---------------------------------------------------------------
```

adutfoy replied:
idem

mbaudin47 commented on python/doc/examples/data_analysis/distribution_fitting/plot_estimate_gev_fremantle.py:
```diff
@@ -1,167 +1,470 @@
-result_rl10_prof = factory.buildReturnLevelProfileLikelihoodEstimator(sample, 10.0)
-zm = result_rl10_prof.getParameter()
-print(f"10 years return level (profile)={zm}")
+# **Return level estimate via the profile log-likelihood function of a stationary GEV**
```
```suggestion
# Return level estimate via the profile log-likelihood function of a stationary GEV
# ---------------------------------------------------------------------------------
```

adutfoy replied:
idem

mbaudin47 commented on python/doc/examples/data_analysis/distribution_fitting/plot_estimate_gev_venice.py:
```diff
@@ -1,153 +1,143 @@
 
 # %%
-# First work on the first column after the year column to retain only one max value per year
+# **Stationary GEV modeling from the annual maximum sea-levels**
```
```suggestion
# Stationary GEV modeling from the annual maximum sea-levels
# ----------------------------------------------------------
```

adutfoy replied:
idem

mbaudin47 commented on python/doc/examples/data_analysis/distribution_fitting/plot_estimate_gev_fremantle.py:
```diff
@@ -1,167 +1,470 @@
+# We want to model this dependence because a slight increase in extreme sea-levels might have
+# a significant impact on the safety of coastal flood defenses.
+#
+# We have define the functional basis for each parameter of the GEV model. Even if we have
+# the possibility to affect a time-varying model to each of the 3 parameters :math:`(\mu, \sigma, \xi)`,
```
```suggestion
# We have define the functional basis for each parameter of the GEV model. Even if we can
# set a time-varying model to each of the 3 parameters :math:`(\mu, \sigma, \xi)`,
```

adutfoy replied:
ok!

mbaudin47 commented on python/doc/examples/data_analysis/distribution_fitting/plot_estimate_gev_fremantle.py:
```diff
@@ -1,167 +1,470 @@
+#
+# where:
+#
+# - the *CenterReduce* method where :math:`c = \dfrac{1}{n} \sum_{i=1}^n t_i` is the mean time stamps
+#   and :math:`d = \sqrt{\dfrac{1}{n} \sum_{i=1}^n (t_i-c)^2}` is the standard deviation of the time stamps;
```
This is duplicated from `GeneralizedExtremeValueFactory.buildTimeVarying()`. So I suggest to remove this and point to the API help page.

adutfoy replied:
You are right... and the text is also in the API help page. But it is really important to normalize the data and readers maight not be aware of the point. I wanted to emphasize that point. i propose to maintain the text.

mbaudin47 commented on python/doc/examples/data_analysis/distribution_fitting/plot_estimate_gev_fremantle.py:
```diff
@@ -1,167 +1,470 @@
+#
+# - the *CenterReduce* method where :math:`c = \dfrac{1}{n} \sum_{i=1}^n t_i` is the mean time stamps
+#   and :math:`d = \sqrt{\dfrac{1}{n} \sum_{i=1}^n (t_i-c)^2}` is the standard deviation of the time stamps;
+# - the *MinMax* method where :math:`c = t_1` is the first time and :math:`d = t_n-t_1` the final time;
```
```suggestion
# - the "MinMax" method where :math:`c = t_1` is the initial time and :math:`d = t_n-t_1` the final time;
```

regislebrun commented on lib/src/Uncertainty/Distribution/GeneralizedExtremeValueFactory.cxx:
```diff
@@ -477,100 +427,74 @@ DistributionFactoryResult GeneralizedExtremeValueFactory::buildRMaximaEstimator(
+  const Scalar bpwm1 = GeneralizedExtremeValueFactoryPWM(sorted, 1);
+  const Scalar bpwm2 = GeneralizedExtremeValueFactoryPWM(sorted, 2);
+  const Scalar kst = (2.0 * bpwm1 - mean) / (3.0 * bpwm2 - mean) - std::log(2.0) / std::log(3.0);
+  const Scalar xi0 = -(7.859 + 2.9554 * kst) * kst;
```
The comment
```
 // pwm for the starting point, see fit.gev function from R mev package
```
needs more details. The probability weighted moments method applied to GEV is detailed in https://www.stat.cmu.edu/technometrics/80-89/VOL-27-03/v2703251.pdf
and page 253 there is a second degree polynomial approximation of the solution of the resulting nonlinear equations.

mbaudin47 commented on python/src/GeneralizedExtremeValueFactory_doc.i.in:
```diff
@@ -11,404 +9,410 @@ DistributionFactory, GeneralizedExtremeValue, FrechetFactory, GumbelFactory, Wei
-    Each record of :math:`R` maxima are sorted in decreasing order
-r : int, default=0 (sets :math:`r=R` to use all maxima)
-    Number of maximas taken into account
+- the *CenterReduce* method where :math:`c = \dfrac{1}{n} \sum_{i=1}^n t_i` is the mean time stamps
+  and :math:`d = \sqrt{\dfrac{1}{n} \sum_{i=1}^n (t_i-c)^2}` is the standard deviation of the time stamps;
```
The layout of the equations could be improved here:

![image](https://github.com/openturns/openturns/assets/31351465/4bc74dd6-e4d3-4e59-9e9a-af07e97615ba)

I suggest to used centered equations:
```suggestion
# - the "CenterReduce" method where
#
# .. math::
# 
#     c = \dfrac{1}{n} \sum_{i=1}^n t_i
#
#   is the mean time stamps and:
#
# .. math:
#
#     d = \sqrt{\dfrac{1}{n} \sum_{i=1}^n (t_i-c)^2}
#
#    is the standard deviation of the time stamps;
```
I do not know the conditions to ensure that the intermediate text stays indented.

adutfoy replied:
ok

mbaudin47 reviewed (commented):
The example is very interesting. Thank you for sharing this. A few cosmetic changes may improve it.

adutfoy commented on python/doc/examples/data_analysis/distribution_fitting/plot_estimate_gev_fremantle.py:
```diff
@@ -1,167 +1,470 @@
 """
```
The GEV examples are scattered among all the other examples and gathering them would be nice for those who want to have  all the examples related to extreme value treatment quickly. But creating a "GEV " section inside the 'Distribution fitting' section would be strange because it would break the balance of the section. We could gather them at the end of the example list: but I did not find the file where the list is specified.....

adutfoy commented on python/doc/examples/data_analysis/distribution_fitting/plot_estimate_gev_fremantle.py:
```diff
@@ -1,167 +1,470 @@
+#
+# - the log-likelihood function,
+# - the profile log-likelihood function.
+#
```
ok

adutfoy commented on python/doc/examples/data_analysis/distribution_fitting/plot_estimate_gev_fremantle.py:
```diff
@@ -1,167 +1,470 @@
 from openturns.usecases import coles
 
 data = coles.Coles().fremantle
 print(data[:5])
```
ok!

adutfoy commented on python/doc/examples/data_analysis/distribution_fitting/plot_estimate_gev_fremantle.py:
```diff
@@ -1,167 +1,470 @@
 # %%
-# Estimate the parameters of the GEV by maximizing the log-likehood
-# and compute the parameter distribution
+# **Stationary GEV modeling via the log-likelihood function**
```
I used some ** to emphasize the section since there are many sections in the text. I think it as a help for the Reader.

adutfoy commented on python/doc/examples/data_analysis/distribution_fitting/plot_estimate_gev_fremantle.py:
```diff
@@ -1,167 +1,470 @@
+# - the probability-probability pot,
+# - the quantile-quantile pot,
+# - the return level plot,
+# - the empirical distribution function.
```
ok

adutfoy commented on python/doc/examples/data_analysis/distribution_fitting/plot_estimate_gev_fremantle.py:
```diff
@@ -1,167 +1,470 @@
-print(f"10 years return level={return_level10}")
-return_level_ci10 = zm10.computeBilateralConfidenceInterval(0.95)
-print(f"CI={return_level_ci10}")
+# **Return level estimate from the estimated stationary GEV**
```
I suggest to maintain the **....
